text/3773-must-use-output.md

```diff
@@ -0,0 +1,124 @@
+- People could misuse it to attempt to enforce soundness of `unsafe` code (which it cannot do)
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+- We could also simply not do this but the potential mistakes it catches are real.
+- The name could be something different and preferably shorter. The name used here was suggested by Josh Triplett and is pretty readable and it can be used before stabilization in the absence of better one. A pretty short and clear `#[must_read]` was also suggested by Kevin Reid.
+- We could write `#[must_use]` on a reference parameter instead. The downside would be that this could be mistaken for saying that the *callee* must use the parameter, rather than the *caller*.
+- Make it a `clippy` lint instead. However not everyone uses `clippy` and the need to communicate which arguments are considered important would diminish its effect. `#[must_use]` is a native rustc lint, and this should be as well, using the same machinery.
+- Try to somehow analyze the code and not require the attribute. This seems hard and it could lead into libraries accidentally triggering warnings in users code if they change the body of the function.
+- Implement a more full feature — e.g. by doing some things in "Future possibilities" section. However, this feature should be useful even without them.
+- Have the attribute on the function instead listing the names of paramters. This could make it nicer to extend to support the "or" relationship described in "Future possibilities".
+
+# Prior art
+[prior-art]: #prior-art
+
+I'm not aware of anything other than the unused write warning and `#[must_use]` attribute which are somewhat related.
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+- Attribute name.
+- Perhaps some limitations are not as difficult as I imagine and they could be lifted right away.
+- Should we emit a warning if a function parameter does not have that attribute but is passed to a function with the attribute to encourage propagation?
+
+# Future possibilities
+[future-possibilities]: #future-possibilities
+
+A custom message could be added just like `#[must_use = "..."]` is already available.
+
+The limitations above might be lifted somehow. For instance I think that it'd be useful to also emit a warning if the reference was obtained from a function parameter that is itself *not* marked `#[must_use_output]`.
+
+Have a way to explain to the compiler how smart pointers work so that this can be used on `Pin` as well.
+
+Have an attribute that can express "or" relationship between parameters and return types. For instance, `Vec::pop` is sometimes used to remove the last element and then pass the vec to something else (e.g. trim `\n` at the end) and sometimes it's used to get the last value and the vec is no longer relevant. Someting like `#[must_use_at_least_one(self, return)]` on function could handle this case.
```

Comment:
This should just be the default behaviour when a function has multiple outputs (e.g. an output parameter as well as a return type). You have to use _at least one,_ not all of them.

Reply:
> This should just be the default behaviour when a function has multiple outputs (e.g. an output parameter as well as a return type) that should be used. You have to use at least one, not all of them.

Then we would need a way to override it in the *other* direction, because there are absolutely functions for which you need to use all the output parameters.

Reply:
Yes, to give a specific example: `core::mem::swap`.

But also for `write` and `read` you should use the returned value specifically. Just using the writer is not enough.